internal/provider/kubernetes/helpers.go

```diff
@@ -61,21 +70,152 @@ func validateParentRefs(ctx context.Context, client client.Client, namespace str
+	// controllerName. This is maintained so that the oldestClass does not change
+	// during reboots.
+	oldestClass *gwapiv1b1.GatewayClass
+}
+
+func (cc *controlledClasses) addMatch(gc *gwapiv1b1.GatewayClass) {
+	cc.matchedClasses = append(cc.matchedClasses, gc)
+
+	switch {
+	case cc.oldestClass == nil:
+		cc.oldestClass = gc
+	case gc.CreationTimestamp.Time.Before(cc.oldestClass.CreationTimestamp.Time):
+		cc.oldestClass = gc
+	case gc.CreationTimestamp.Time.Equal(cc.oldestClass.CreationTimestamp.Time) && gc.Name < cc.oldestClass.Name:
+		// tie-breaker: first one in alphabetical order is considered oldest/accepted
+		cc.oldestClass = gc
+	}
+}
+
+func (cc *controlledClasses) removeMatch(gc *gwapiv1b1.GatewayClass) {
+	// First remove gc from matchedClasses.
+	for i, matchedGC := range cc.matchedClasses {
+		if matchedGC.Name == gc.Name {
+			cc.matchedClasses[i] = cc.matchedClasses[len(cc.matchedClasses)-1]
+			cc.matchedClasses = cc.matchedClasses[:len(cc.matchedClasses)-1]
+			break
+		}
+	}
+
+	// If the oldestClass is removed, find the new oldestClass candidate
+	// from matchedClasses.
+	if cc.oldestClass != nil && cc.oldestClass.Name == gc.Name {
+		if len(cc.matchedClasses) == 0 {
+			cc.oldestClass = nil
+			return
+		}
+
+		cc.oldestClass = cc.matchedClasses[0]
+		for i := 1; i < len(cc.matchedClasses); i++ {
+			current := cc.matchedClasses[i]
+			if current.CreationTimestamp.Time.Before(cc.oldestClass.CreationTimestamp.Time) ||
+				(current.CreationTimestamp.Time.Equal(cc.oldestClass.CreationTimestamp.Time) &&
+					current.Name < cc.oldestClass.Name) {
+				cc.oldestClass = current
+				return
+			}
+		}
 	}
+}
 
-	httpRouteList := &gwapiv1b1.HTTPRouteList{}
-	if err := c.List(ctx, httpRouteList, &client.ListOptions{Namespace: ns}); err != nil {
-		return false, fmt.Errorf("error listing httproutes")
+func (cc *controlledClasses) acceptedClass() *gwapiv1b1.GatewayClass {
+	return cc.oldestClass
+}
+
+func (cc *controlledClasses) notAcceptedClasses() []*gwapiv1b1.GatewayClass {
+	var res []*gwapiv1b1.GatewayClass
+	for _, gc := range cc.matchedClasses {
+		// skip the oldest one since it will be accepted.
+		if gc.Name != cc.oldestClass.Name {
+			res = append(res, gc)
+		}
 	}
 
-	if len(tlsRouteList.Items)+len(httpRouteList.Items) > 0 {
-		return true, nil
+	return res
+}
+
+// isAccepted returns true if the provided gatewayclass contains the Accepted=true
+// status condition.
+func isAccepted(gc *gwapiv1b1.GatewayClass) bool {
+	if gc == nil {
+		return false
 	}
-	return false, nil
+	for _, cond := range gc.Status.Conditions {
+		if cond.Type == string(gwapiv1b1.GatewayClassConditionStatusAccepted) && cond.Status == metav1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+// gatewaysOfClass returns a list of gateways that reference gc from the provided gwList.
+func gatewaysOfClass(gc *gwapiv1b1.GatewayClass, gwList *gwapiv1b1.GatewayList) []gwapiv1b1.Gateway {
```

Comment:
Many of these helper funcs had unit tests in the {controller_name}_test.go. Can you either add these tests or create an issue and link to this PR so they can be readded in a separate PR?